File: extension/src/dialer/turn_creds.js
// Client-side TURN credential derivation (PR #42.3, JS mirror of
// `openhost-client::turn_creds`). The password is a public function
// of the daemon's Ed25519 identity — anyone with the `oh://<pk>/`
// URL can compute it. TURN long-term auth needs MESSAGE-INTEGRITY
// HMAC with matching inputs on both peers; this provides that
// matching input without a shared secret.

export const TURN_REALM = "openhost";
export const TURN_USERNAME = "openhost";

// Decode the 52-char z-base-32 daemon pubkey to the raw 32 Ed25519 bytes.
// Minimal zbase32 decoder — the alphabet is openhost's 52-char canonical
// form "ybndrfg8ejkmcpqxot1uwisza345h769".
const ZBASE32_ALPHABET = "ybndrfg8ejkmcpqxot1uwisza345h769";

function zbase32Decode(s) {
  const bits = [];
  for (const ch of s.toLowerCase()) {
    const idx = ZBASE32_ALPHABET.indexOf(ch);
    if (idx < 0) throw new Error(`invalid zbase32 char: ${ch}`);
    for (let i = 4; i >= 0; i--) bits.push((idx >> i) & 1);
  }
  // We encoded 32 bytes = 256 bits into 52 chars * 5 = 260 bits;
  // the trailing 4 bits are padding zeros — drop them.
  const trimmed = bits.slice(0, 256);
  const out = new Uint8Array(32);
  for (let i = 0; i < 32; i++) {
    let byte = 0;
    for (let j = 0; j < 8; j++) byte = (byte << 1) | trimmed[i * 8 + j];
    out[i] = byte;
  }
  return out;
}

function hexEncode(bytes) {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}

// Match `openhost-daemon::turn_server::password_for_daemon` byte-for-byte:
// `lower_hex(sha256("openhost-turn-v1" || daemon_pk))[..32]`.
export async function passwordForDaemon(daemonPkZbase32) {
  const pkBytes = zbase32Decode(daemonPkZbase32);
  const prefix = new TextEncoder().encode("openhost-turn-v1");
  const msg = new Uint8Array(prefix.length + pkBytes.length);
  msg.set(prefix, 0);
  msg.set(pkBytes, prefix.length);
  const digest = await crypto.subtle.digest("SHA-256", msg);
  return hexEncode(new Uint8Array(digest).slice(0, 16));
}

// Build the `RTCIceServer` entry to append to the `iceServers` list
// when the resolved host record carries a v3 `turn_endpoint`. Returns
// `null` if no TURN endpoint was advertised.
export async function turnIceServerFor(daemonPkZbase32, turnEndpoint) {
  if (!turnEndpoint) return null;
  const password = await passwordForDaemon(daemonPkZbase32);
  return {
    urls: [`turn:${turnEndpoint.ip}:${turnEndpoint.port}`],
    username: TURN_USERNAME,
    credential: password,
  };
}

File: crates/openhost-client/src/turn_creds.rs
//! Client-side TURN credential derivation (PR #42.3).
//!
//! Mirrors `openhost-daemon::turn_server` — both sides compute the
//! same password from the daemon's public Ed25519 identity, so a
//! client with only the `oh://<daemon-pk>/` URL can authenticate
//! against the daemon's embedded TURN relay without any pre-shared
//! secret. The password is not actually secret (anyone who knows the
//! daemon pubkey can compute it), but TURN long-term auth requires
//! MESSAGE-INTEGRITY HMAC with matching inputs on both peers; this
//! function provides that matching input.

use openhost_core::identity::PublicKey;

/// Fixed TURN realm advertised by openhost daemons.
pub const TURN_REALM: &str = "openhost";

/// Fixed TURN username clients must present.
pub const TURN_USERNAME: &str = "openhost";

/// Compute the long-term TURN password for a daemon, from its public
/// key. See `openhost-daemon::turn_server::password_for_daemon` — the
/// two functions MUST agree byte-for-byte.
pub fn password_for_daemon(daemon_pk: &PublicKey) -> String {
    use sha2::{Digest, Sha256};
    let mut h = Sha256::new();
    h.update(b"openhost-turn-v1");
    h.update(daemon_pk.to_bytes());
    let digest = h.finalize();
    hex::encode(&digest[..16])
}

#[cfg(test)]
mod tests {
    use super::*;
    use openhost_core::identity::SigningKey;

    #[test]
    fn deterministic_per_daemon() {
        let sk = SigningKey::from_bytes(&[0x11u8; 32]);
        let pk = sk.public_key();
        assert_eq!(password_for_daemon(&pk), password_for_daemon(&pk));
    }

    #[test]
    fn differs_across_daemons() {
        let sk_a = SigningKey::from_bytes(&[0x11u8; 32]);
        let sk_b = SigningKey::from_bytes(&[0x22u8; 32]);
        assert_ne!(
            password_for_daemon(&sk_a.public_key()),
            password_for_daemon(&sk_b.public_key())
        );
    }
}
